src/absorptionlines.h

```diff
@@ -0,0 +1,961 @@
+  std::vector<Rational> mupperquanta;
+
+public:
+  /** Default initialization 
+   * 
+   * @param[in] F0 Central frequency
+   * @param[in] I0 Reference line strength at external T0
+   * @param[in] E0 Lower energy level
+   * @param[in] glow Lower level statistical weight
+   * @param[in] gupp Upper level statistical weight
+   * @param[in] A Einstein spontaneous emission coefficient
+   * @param[in] zeeman Zeeman model
+   * @param[in] lineshape Line shape model
+   * @param[in] lowerquanta Lower quantum numbers
+   * @param[in] upperquanta Upper quantum numbers
+   */
+  SingleLine(Numeric F0=0,
+             Numeric I0=0,
+             Numeric E0=0,
+             Numeric glow=0,
+             Numeric gupp=0,
+             Numeric A=0,
+             Zeeman::Model zeeman=Zeeman::Model(),
+             const LineShape::Model2& lineshape=LineShape::Model2(),
+             const std::vector<Rational>& lowerquanta={},
+             const std::vector<Rational>& upperquanta={}) :
+             mF0(F0),
+             mI0(I0),
+             mE0(E0),
+             mglow(glow),
+             mgupp(gupp),
+             mA(A),
+             mzeeman(zeeman),
+             mlineshape(lineshape),
+             mlowerquanta(lowerquanta),
+             mupperquanta(upperquanta) {}
+  
+  /** Initialization for constant sizes
+   * 
+   * @param[in] nbroadeners Number of broadening species
+   * @param[in] nquanta Number of local quantum numbers
+   */
+  SingleLine(size_t nbroadeners, size_t nquanta) noexcept :
+  mlineshape(nbroadeners), mlowerquanta(nquanta), mupperquanta(nquanta) {}
+  
+  //////////////////////////////////////////////////////////////////
+  /////////////////////////////////////////////////////////// Counts
+  //////////////////////////////////////////////////////////////////
+  
+  /** Number of lineshape elements */
+  Index LineShapeElems() const noexcept {return mlineshape.nelem();}
+  
+  /** Number of lower quantum numbers */
+  Index LowerQuantumElems() const noexcept {return mlowerquanta.size();}
+  
+  /** Number of upper quantum numbers */
+  Index UpperQuantumElems() const noexcept {return mupperquanta.size();}
+  
+  //////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////// Constant access
+  //////////////////////////////////////////////////////////////////
+  
+  /** Central frequency */
+  Numeric F0() const noexcept {return mF0;}
+  
+  /** Lower level energy */
+  Numeric E0() const noexcept {return mE0;}
+  
+  /** Reference line strength */
+  Numeric I0() const noexcept {return mI0;}
+  
+  /** Einstein spontaneous emission */
+  Numeric A() const noexcept {return mA;}
+  
+  /** Lower level statistical weight */
+  Numeric g_low() const noexcept {return mglow;}
+  
+  /** Upper level statistical weight */
+  Numeric g_upp() const noexcept {return mgupp;}
+  
+  /** Zeeman model */
+  auto Zeeman() const noexcept {return mzeeman;}

```

Comment:
I am weighing in here, because I have strong feelings about the auto keyword. In my opinion it should generally be used only for lambdas and to avoid complicated type specifications involving the typename keyword.

As a user of this class, I should not be required to know the types of private class member to understand the signature of public functions. Also having auto as return type in the public interface of a class is counterproductive as it weakens the guarantees that this interface is meant to provide. So this basically decreases modularity and encapsulation of the code base which I consider quite fundamental principles. 

Imagine you want to at some point change the type of mzeeman? Then suddenly ARTS breaks everywhere the Zeeman function is used.

Reply:
It is the complete opposite.

The auto keyword means that any changes in the Zeeman class must guarantee that the interfaces of the class is the same.  Even if you move the class to a different module or namespace.  The auto keyword means the implementation is just a detail.

Besides, if the interface to a class has to change in anyway that matters, then ARTS will break everywhere that it is used anyways.  Also, any decent IDE will show you the class anyways.  If it does not, you need to report the bug.

Reply:
Regarding your first argument: This only solves the problem of the interface changing due to implementation changes. If auto would not be used here, the interface wouldn't have changed in the first place.

Regarding your second argument: I am not talking about interface changes. I am talking about implementation changes. The point of interfaces is that the implementation can change without affecting other users code. This is why you have private and public visibility: To avoid that users become dependent on code that they are not meant to.

So my initial argument still stands: The problem is that with the auto keyword you tie the interface to the implementation. Yes, this is checked at compile time but it will still require you to mess with other peoples code when you want to change the implementation of your class. This is not the case if you have a clear separation of interface and implementation.

Reply:
The implementation is a detail that is not very important since it can change, so I cannot follow your argument.

The auto keyword is parts of C++ standard because it makes coding easier.  `for (auto& x: y)` is more easy to write and read than `for (const LineShape::SingleSpeciesModel& x: y)`, `constexpr auto x = "ab";` is more readable than `constexpr const char * const x = "ab";`, and `auto Zeeman() const noexcept {return mzeeman;}` is more readable than `Zeeman::Model Zeeman() const noexcept {return mzeeman;}`.  It is such a common pattern to have return variables that are just passed around that auto simplifies things tremendously.  But you want all of the above cases to use the long-form instead?

Reply:
Yes, because writing out return types of functions is nothing that needs to be simplified. It's just not a critical difficulty in software design. The critical difficulty, at least in ARTS, is managing complexity. And encapsulation and modularity have proven effective in dealing with complexity. This is basically why we are using an object oriented programming language.

Anyways, we do not seem to get anywhere with this discussion. Maybe I am too conservative with regards to new language features. This was meant as advice and you should of course feel free to ignore it.

Reply:
I appreciate your input.  I still cannot follow your argument.  IMO, the auto-keyword just limits the number of places you have to touch the code when you make otherwise breaking changes.  

In your's and Oliver's opinions, it makes the code less readable.  So in practice, I will have to change this to non-auto variables.  You guys are the official maintainers, so I should follow the new coding style limitations.  I am now merely wondering in how many places I should go through the code to follow this official coding style.  Because I have used stuff like `for(auto& x: y)` for arrays  and `auto x = f()` for annoyingly long variable names all over the place.  So there will be quite a lot of changes to remove these things.

Reply:
> The implementation is a detail that is not very important since it can change, so I cannot follow your argument.

@riclarsson It is not an implementation detail anymore if the member variable is directly exposed by an access method in the interface of the class. Since this class is probably used in other parts of ARTS than your own, other developers need to interact with it. How does someone who wants to call member functions of your class know what type the functions return if the interface just says `auto`? You're forcing users of your class to look at the implementation, find which member is returned and look up its type. Probably even going to the .cc file if the method is not implemented right in the header. This might be okay for yourself because you know the implementation of your methods, but it is making it unnecessarily more tedious for your fellow developers to use your code.

> The auto keyword is parts of C++ standard because it makes coding easier. `for (auto& x: y)` is more easy to write and read than `for (const LineShape::SingleSpeciesModel& x: y)`, `constexpr auto x = "ab";` is more readable than `constexpr const char * const x = "ab";`, and `auto Zeeman() const noexcept {return mzeeman;}` is more readable than `Zeeman::Model Zeeman() const noexcept {return mzeeman;}`. It is such a common pattern to have return variables that are just passed around that auto simplifies things tremendously. But you want all of the above cases to use the long-form instead?

Just because the feature exists doesn't mean it's a good idea to use it everywhere and whenever possible. Simon and me were only objecting using `auto` in class interfaces, nobody claimed you should never use `auto`. The loop is a perfect example where it makes sense to use `auto` because it makes the code more readable. And in this case it is actually an implementation detail, because it doesn't have any impact for callers of your method how you implement the loop. The variable definition is also a good example because one can see what is assigned right there, writing out the type would be redundant.

The purpose of class and method declarations is to establish a clear protocol between you and the user of your class on how to communicate with each other. With a return type of `auto`, your interface basically tells me: If you call me, you get something, but I won't tell you what.

Reply:
Right, this is silly.  You have made a decision to disallow all autos except in lambdas.  I will shortly commit the code.  It will be worse to read but follow your new style demands.

Reply:
> I appreciate your input. I still cannot follow your argument. IMO, the auto-keyword just limits the number of places you have to touch the code when you make otherwise breaking changes.

Wouldn't `auto` only help in the case that a type returned by a function is renamed, but its implementation stays exactly the same? Otherwise code calling the changed method most likely needs to be adapted anyway. Saving refactoring work is not a justification to use `auto` in interfaces since the refactoring feature of IDEs can take care of automatically renaming all occurrences of a type name in the whole project.

The priority of an interface should be focused on conveying as much information as possible on how to interact with the class, not on how convenient it is for the maintainer to refactor it. The former should not be sacrificed for the latter.

> In your's and Oliver's opinions, it makes the code less readable. So in practice, I will have to change this to non-auto variables. You guys are the official maintainers, so I should follow the new coding style limitations. I am now merely wondering in how many places I should go through the code to follow this official coding style. Because I have used stuff like `for(auto& x: y)` for arrays and `auto x = f()` for annoyingly long variable names all over the place. So there will be quite a lot of changes to remove these things.

Just to be clear again: Nobody is suggesting to not use `auto` in such implementations. We're only talking about class/function declarations.

Reply:
This is false Oliver.  I asked "But you want all of the above cases to use the long-form instead?" and the response I got in short form reads "Yes".

Reply:
Sorry if I wrote anything that led to this misunderstanding. Can you please quote the part that gave you the impression that I want all autos removed?

Reply:
Not you, you only commented on the implementation of this code. Simon did have more general rules to follow in the conversation above. You guys are the maintainers so it is silly if I do not comply.

Reply:
Ok, I see which post from Simon you were referring to. Although it was not just a "Yes.". He wrote 'Yes, because writing out return types of functions is nothing that needs to be simplified.' Therefore I assumed that his 'Yes' was limited to the return types and did not apply to the for loops.

Reply:
Let's postpone this till Monday. I think @simonpf and me also need to iterate on that again to make a consistent decision. Sorry again for the confusion.

Reply:
OK, sorry, I see that my answer caused some misunderstanding. For me this discussion was solely about the use of auto in public interfaces of classes. And this is the usage that I clearly oppose.

For the rest, I think, there's not really a point in making strict rules as I see this being more related to readability and code preferences and less to code quality.

Reply:
Right, my bad.  I will delete the commit and then follow your interface demands.  I still protest that this is wasteful and makes it easier to introduce bugs.  But if your development environment cannot show the variable type on request, I can see this be an issue.